File: typescript/src/bc/workflow.ts
/**
 * Workflow composition: the dynamic `run-steps` engine, composite pipelines,
 * argument parsing and parameter helpers. Ported from big-config.workflow.
 */
import {
  type Opts,
  type StepFn,
  type WfStep,
  ok,
  toStepFn,
} from "./core.js";
import { toWorkflowStar } from "./pluggable.js";
import { renderTemplates } from "./render.js";
import { runCmds } from "./run.js";
import { keywordToName, keywordToPath } from "./utils.js";

/** Logs the command about to run, to stderr. */
export const printStepFn: StepFn = toStepFn({
  beforeF: (step, opts) => {
    if (step === "big-config.run/run-cmd" && opts.cmds?.[0]) {
      console.error(`> ${opts.cmds[0]}`);
    }
  },
});

const DEFAULT_GLOBALS = [
  "env",
  "shellOpts",
  "module",
  "profile",
  "prefix",
  "objectPrefix",
  "globals",
];

/** Copy the global options across workflows. */
export function selectGlobals(opts: Opts): Opts {
  const keys: string[] = opts.globals ?? DEFAULT_GLOBALS;
  const out: Opts = {};
  for (const k of keys) {
    if (k in opts) out[k] = opts[k];
  }
  return out;
}

function resolveFn(opts: Opts, key: string, dflt?: unknown): any {
  const f = opts[key];
  if (f == null) {
    if (dflt === undefined) throw new Error(`\`${key}\` not defined`);
    return dflt;
  }
  return f;
}

/**
 * The dynamic "workflow of workflows": runs the steps listed under `opts.steps`.
 * `create` / `delete` are entire subworkflows; the rest are ordinary steps.
 */
export function runSteps(stepFns: StepFn[], opts: Opts): Opts {
  const globalsOpts = selectGlobals(opts);
  const createOpts = { ...(opts.createOpts ?? {}), ...globalsOpts };
  const deleteOpts = { ...(opts.deleteOpts ?? {}), ...globalsOpts };
  let cur: Opts = { ...opts };
  const queue: string[] = (opts.steps ?? []).map((s: string) =>
    s.includes("/") ? s : `big-config.workflow/${s}`,
  );
  const twoArgOk = (_sf: StepFn[], o: Opts) => ok(o);
  const wf = toWorkflowStar({
    firstStep: "big-config.workflow/start",
    lastStep: "big-config.workflow/end",
    wireFn: (step, sfns): [WfStep, string | null] => {
      switch (step) {
        case "big-config.workflow/start":
          return [ok, null];
        case "big-config.workflow/render":
          return [(o) => renderTemplates(sfns, o), null];
        case "big-config.workflow/exec":
          return [(o) => runCmds(sfns, o), null];
        case "big-config.workflow/create":
          return [(o) => resolveFn(opts, "createFn")(sfns, o), null];
        case "big-config.workflow/delete":
          return [(o) => resolveFn(opts, "deleteFn")(sfns, o), null];
        case "big-config.workflow/validate":
          return [(o) => resolveFn(opts, "validateFn", twoArgOk)(sfns, o), null];
        case "big-config.workflow/describe":
          return [(o) => resolveFn(opts, "describeFn", twoArgOk)(sfns, o), null];
        default:
          return [(o) => o, null];
      }
    },
    nextFn: (step, _nextStep, o) => {
      if (
        step === "big-config.workflow/create" ||
        step === "big-config.workflow/delete"
      ) {
        cur = { ...cur, exit: o.exit, err: o.err, [step]: [...(cur[step] ?? []), o] };
      } else {
        cur = o;
      }
      if (step === "big-config.workflow/end") return [null, cur];
      if (typeof o.exit === "number" && o.exit > 0) {
        return ["big-config.workflow/end", cur];
      }
      const next = queue.shift();
      if (next) {
        const stepOpts =
          next === "big-config.workflow/create"
            ? createOpts
            : next === "big-config.workflow/delete"
              ? deleteOpts
              : cur;
        return [next, stepOpts];
      }
      return ["big-config.workflow/end", cur];
    },
  });
  return wf(stepFns, cur);
}

const PARSE_ARGS_STEPS = new Set([
  "lock",
  "git-check",
  "render",
  "create",
  "delete",
  "validate",
  "describe",
  "exec",
  "git-push",
  "unlock-any",
]);

/** Normalize a string or array of CLI args into `{ steps, cmds }`. */
export function parseArgs(input: string | string[]): {
  steps: string[];
  cmds: string[];
} {
  const xs: string[] =
    typeof input === "string"
      ? input.trim().split(/\s+/).filter(Boolean)
      : [...input];
  const steps: string[] = [];
  const cmds: string[] = [];
  let i = 0;
  while (i < xs.length) {
    const token = xs[i];
    if (PARSE_ARGS_STEPS.has(token)) {
      steps.push(token);
      i++;
      continue;
    }
    if (token === "--") {
      const rest = xs.slice(i + 1);
      if (rest.length === 0) {
        throw new Error("-- cannot be without a command");
      }
      if (!steps.includes("exec")) steps.push("exec");
      cmds.push(rest.join(" "));
      break;
    }
    if (!steps.includes("exec")) steps.push("exec");
    cmds.push(token.replace(/:/g, " "));
    i++;
  }
  return { steps, cmds };
}

function hashHex(s: string): string {
  let h = 0;
  for (let i = 0; i < s.length; i++) {
    h = (Math.imul(31, h) + s.charCodeAt(i)) | 0;
  }
  return (h >>> 0).toString(16);
}

/**
 * Derive a deterministic `prefix` / `objectPrefix` from `firstStep` and the
 * profile, so a workflow always resolves to the same directory.
 */
export function newPrefix(opts: Opts, firstStep: string): Opts {
  const prefix: string = opts.prefix ?? ".dist";
  const objectPrefix: string = opts.objectPrefix ?? "tofu";
  const profile: string = opts.profile ?? "default";
  const dirs = prefix.split("/");
  const objectDirs = objectPrefix.split("/");
  const last = dirs[dirs.length - 1];
  const profileFound = last.startsWith(profile);
  const prevHash = profileFound ? (last.split("-").pop() ?? "") : "";
  const baseDirs = profileFound ? dirs.slice(0, -1) : dirs;
  const baseObjectDirs = profileFound ? objectDirs.slice(0, -1) : objectDirs;
  const suffix = hashHex(firstStep + prevHash);
  return {
    ...opts,
    prefix: [...baseDirs, `${profile}-${suffix}`].join("/"),
    objectPrefix: [...baseObjectDirs, `${profile}-${suffix}`].join("/"),
  };
}

/** Resolve the directory of a previous workflow, to extract its outputs. */
export function path(opts: Opts, name: string): string {
  return `${opts.prefix ?? ".dist"}/${keywordToPath(name)}`;
}

/**
 * Prepare `opts`: compute target directory / object, merge params into every
 * template, and set the shell working directory.
 */
export function prepare(base: Opts, overrides: Opts): Opts {
  const name = base.name;
  if (name == null) throw new Error("Argument name is nil");
  if (overrides == null) throw new Error("Argument overrides is nil");
  const merged = { ...base, ...overrides };
  const prefix: string | undefined = merged.prefix;
  const objectPrefix: string | undefined = merged.objectPrefix;
  const pathFn: (o: Opts) => string =
    merged.pathFn ?? ((o) => `${prefix ?? ".dist"}/${keywordToPath(o.name)}`);
  const objectFn: (o: Opts) => string =
    merged.objectFn ??
    ((o) => `${objectPrefix ?? "tofu"}/${keywordToName(o.name)}`);
  const dir = pathFn(merged);
  const object = objectFn(merged);
  const params: Record<string, any> = merged.params ?? {};
  const templates = (merged.templates ?? []).map((t: any) => ({
    ...t,
    ...params,
    targetDir: dir,
    "target-object": object,
  }));
  return {
    ...merged,
    templates,
    shellOpts: { ...(merged.shellOpts ?? {}), dir },
  };
}

/** Merge package params into per-tool create/delete opts. */
export function mergeParams(
  tools: string[],
  params: Record<string, any>,
  opts: Opts,
): Opts {
  let result: Opts = { ...opts };
  for (const tool of tools) {
    for (const slot of ["createOpts", "deleteOpts"]) {
      const toolOpts = result[slot]?.[tool];
      if (toolOpts?.params) {
        result = {
          ...result,
          [slot]: {
            ...result[slot],
            [tool]: { ...toolOpts, params: { ...params, ...toolOpts.params } },
          },
        };
      }
    }
  }
  return result;
}

const BC_PAR_PREFIX = "BC_PAR_";

/** Override params from `BC_PAR_*` environment variables. */
export function readBcPars(
  opts: Opts,
  env: NodeJS.ProcessEnv = process.env,
): Opts {
  const fromEnv: Record<string, string> = {};
  for (const [k, v] of Object.entries(env)) {
    if (k.startsWith(BC_PAR_PREFIX) && v !== undefined) {
      const key = k
        .slice(BC_PAR_PREFIX.length)
        .toLowerCase()
        .replace(/_/g, "-")
        .replace(/\./g, "-");
      fromEnv[key] = v;
    }
  }
  return { ...opts, params: { ...(opts.params ?? {}), ...fromEnv } };
}

export interface PipelineEntry {
  /** Step keyword, e.g. `tools/tofu`. Used to store per-step opts/results. */
  name: string;
  /** The tool workflow function. */
  fn: (stepFns: StepFn[], opts: Opts) => Opts;
  /** CLI-style arguments, e.g. `render tofu:init tofu:apply:-auto-approve`. */
  args: string;
  /** Adapts outputs from previous steps into this step's params. */
  optsFn?: (opts: Opts) => Opts;
}

/** Create a composite workflow from a pipeline of tool workflows. */
export function toCompWorkflow({
  firstStep,
  lastStep,
  pipeline,
}: {
  firstStep: string;
  lastStep?: string;
  pipeline: PipelineEntry[];
}): (stepFns: StepFn[], opts: Opts) => Opts {
  return (stepFns, opts) => {
    const last = lastStep ?? `${firstStep.split("/")[0]}/end`;
    const globalsOpts = newPrefix(selectGlobals(opts), firstStep);
    const stepInfo = new Map<
      string,
      { stepOpts: Opts; optsFn: (o: Opts) => Opts }
    >();
    for (const entry of pipeline) {
      const { steps, cmds } = parseArgs(entry.args);
      const override = opts[`${entry.name}-opts`] ?? {};
      stepInfo.set(entry.name, {
        stepOpts: { steps, cmds, ...globalsOpts, ...override },
        optsFn: entry.optsFn ?? ((o) => o),
      });
    }
    const order = [firstStep, ...pipeline.map((e) => e.name), last];
    const nextByStep = new Map<string, string>();
    for (let i = 0; i < order.length - 1; i++) {
      nextByStep.set(order[i], order[i + 1]);
    }
    const fnByStep = new Map<string, WfStep>();
    fnByStep.set(firstStep, ok);
    fnByStep.set(last, (o) => o);
    for (const e of pipeline) {
      fnByStep.set(e.name, (o) => e.fn(stepFns, o));
    }
    const stepsSet = new Set(pipeline.map((e) => e.name));
    let cur: Opts = { ...opts };
    const wf = toWorkflowStar({
      firstStep,
      lastStep: last,
      wireFn: (step): [WfStep, string | null] => [
        fnByStep.get(step) ?? ((o) => o),
        nextByStep.get(step) ?? null,
      ],
      nextFn: (step, nextStep, o) => {
        if (stepsSet.has(step)) {
          cur = { ...cur, exit: o.exit, err: o.err, [step]: o };
        } else {
          cur = o;
        }
        if (step === last) return [null, cur];
        if (typeof o.exit === "number" && o.exit > 0) return [last, cur];
        const info = nextStep ? stepInfo.get(nextStep) : undefined;
        if (info) return [nextStep, info.optsFn(info.stepOpts)];
        return [nextStep, cur];
      },
    });
    return wf(stepFns, opts);
  };
}

File: src/once/bc/workflow.py
"""Workflow composition, dynamic step runner, arg parsing, and params helpers."""
from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from typing import Any, Callable

from .core import Opts, StepFn, WfStep, ok, to_step_fn, to_workflow
from .pluggable import to_workflow_star
from .render import render_templates
from .run import run_cmds
from .utils import keyword_to_name, keyword_to_path


def _print_step_before(step: str, opts: Opts) -> None:
    if step == "big-config.run/run-cmd" and (opts.get("cmds") or [None])[0]:
        print(f"> {(opts.get('cmds') or [''])[0]}", file=sys.stderr)


print_step_fn: StepFn = to_step_fn(before_f=_print_step_before)

DEFAULT_GLOBALS = ["env", "shellOpts", "module", "profile", "prefix", "objectPrefix", "globals"]


def select_globals(opts: Opts) -> Opts:
    """Copy global options across workflows."""
    keys = opts.get("globals") or DEFAULT_GLOBALS
    return {k: opts[k] for k in keys if k in opts}


def _resolve_fn(opts: Opts, key: str, default: Any = None, has_default: bool = False) -> Any:
    f = opts.get(key)
    if f is None:
        if not has_default:
            raise KeyError(f"`{key}` not defined")
        return default
    return f


def run_steps(step_fns: list[StepFn], opts: Opts) -> Opts:
    """Dynamic workflow of workflows: run steps listed under ``opts['steps']``."""
    globals_opts = select_globals(opts)
    create_opts = {**(opts.get("createOpts") or {}), **globals_opts}
    delete_opts = {**(opts.get("deleteOpts") or {}), **globals_opts}
    cur: Opts = dict(opts)
    queue = [s if "/" in s else f"big-config.workflow/{s}" for s in (opts.get("steps") or [])]

    def two_arg_ok(_sf: list[StepFn], o: Opts) -> Opts:
        return ok(o)

    def wire(step: str, sfns: list[StepFn]) -> tuple[WfStep, str | None]:
        if step == "big-config.workflow/start":
            return ok, None
        if step == "big-config.workflow/render":
            return lambda o: render_templates(sfns, o), None
        if step == "big-config.workflow/exec":
            return lambda o: run_cmds(sfns, o), None
        if step == "big-config.workflow/create":
            return lambda o: _resolve_fn(opts, "createFn")(sfns, o), None
        if step == "big-config.workflow/delete":
            return lambda o: _resolve_fn(opts, "deleteFn")(sfns, o), None
        if step == "big-config.workflow/validate":
            return lambda o: _resolve_fn(opts, "validateFn", two_arg_ok, True)(sfns, o), None
        if step == "big-config.workflow/describe":
            return lambda o: _resolve_fn(opts, "describeFn", two_arg_ok, True)(sfns, o), None
        return lambda o: o, None

    def next_fn(step: str, _next_step: str | None, o: Opts) -> tuple[str | None, Opts]:
        nonlocal cur
        if step in {"big-config.workflow/create", "big-config.workflow/delete"}:
            cur = {**cur, "exit": o.get("exit"), "err": o.get("err"), step: [*(cur.get(step) or []), o]}
        else:
            cur = o
        if step == "big-config.workflow/end":
            return None, cur
        if isinstance(o.get("exit"), int) and o["exit"] > 0:
            return "big-config.workflow/end", cur
        if queue:
            nxt = queue.pop(0)
            if nxt == "big-config.workflow/create":
                step_opts = create_opts
            elif nxt == "big-config.workflow/delete":
                step_opts = delete_opts
            else:
                step_opts = cur
            return nxt, step_opts
        return "big-config.workflow/end", cur

    wf = to_workflow_star(
        first_step="big-config.workflow/start",
        last_step="big-config.workflow/end",
        wire_fn=wire,
        next_fn=next_fn,
    )
    return wf(step_fns, cur)


PARSE_ARGS_STEPS = {
    "lock",
    "git-check",
    "render",
    "create",
    "delete",
    "validate",
    "describe",
    "exec",
    "git-push",
    "unlock-any",
}


def parse_args(input_: str | list[str] | tuple[str, ...]) -> dict[str, list[str]]:
    """Normalize CLI args into ``{'steps': [...], 'cmds': [...]}``."""
    if isinstance(input_, str):
        xs = [x for x in input_.strip().split() if x]
    else:
        xs = list(input_)
    steps: list[str] = []
    cmds: list[str] = []
    i = 0
    while i < len(xs):
        token = xs[i]
        if token in PARSE_ARGS_STEPS:
            steps.append(token)
            i += 1
            continue
        if token == "--":
            rest = xs[i + 1 :]
            if not rest:
                raise ValueError("-- cannot be without a command")
            if "exec" not in steps:
                steps.append("exec")
            cmds.append(" ".join(rest))
            break
        if "exec" not in steps:
            steps.append("exec")
        cmds.append(token.replace(":", " "))
        i += 1
    return {"steps": steps, "cmds": cmds}


def _hash_hex(s: str) -> str:
    h = 0
    for ch in s:
        h = (31 * h + ord(ch)) & 0xFFFFFFFF
    return format(h, "x")


def new_prefix(opts: Opts, first_step: str) -> Opts:
    """Derive deterministic ``prefix`` / ``objectPrefix`` from the profile."""
    prefix = opts.get("prefix") or ".dist"
    object_prefix = opts.get("objectPrefix") or "tofu"
    profile = opts.get("profile") or "default"
    dirs = str(prefix).split("/")
    object_dirs = str(object_prefix).split("/")
    last = dirs[-1]
    profile_found = last.startswith(str(profile))
    prev_hash = last.split("-")[-1] if profile_found else ""
    base_dirs = dirs[:-1] if profile_found else dirs
    base_object_dirs = object_dirs[:-1] if profile_found else object_dirs
    suffix = _hash_hex(first_step + prev_hash)
    return {
        **opts,
        "prefix": "/".join([*base_dirs, f"{profile}-{suffix}"]),
        "objectPrefix": "/".join([*base_object_dirs, f"{profile}-{suffix}"]),
    }


def path(opts: Opts, name: str) -> str:
    """Resolve the directory of a previous workflow."""
    return f"{opts.get('prefix') or '.dist'}/{keyword_to_path(name)}"


def prepare(base: Opts, overrides: Opts) -> Opts:
    """Compute target directories, merge params into templates, and set cwd."""
    name = base.get("name")
    if name is None:
        raise ValueError("Argument name is nil")
    if overrides is None:
        raise ValueError("Argument overrides is nil")
    merged = {**base, **overrides}
    prefix = merged.get("prefix")
    object_prefix = merged.get("objectPrefix")
    path_fn = merged.get("pathFn") or (lambda o: f"{prefix or '.dist'}/{keyword_to_path(o['name'])}")
    object_fn = merged.get("objectFn") or (lambda o: f"{object_prefix or 'tofu'}/{keyword_to_name(o['name'])}")
    directory = path_fn(merged)
    obj = object_fn(merged)
    params = merged.get("params") or {}
    templates = [
        {**t, **params, "targetDir": directory, "target-object": obj}
        for t in (merged.get("templates") or [])
    ]
    return {**merged, "templates": templates, "shellOpts": {**(merged.get("shellOpts") or {}), "dir": directory}}


def merge_params(tools: list[str], params: dict[str, Any], opts: Opts) -> Opts:
    """Merge package params into per-tool create/delete opts."""
    result = dict(opts)
    for tool in tools:
        for slot in ["createOpts", "deleteOpts"]:
            tool_opts = (result.get(slot) or {}).get(tool)
            if isinstance(tool_opts, dict) and isinstance(tool_opts.get("params"), dict):
                result = {
                    **result,
                    slot: {
                        **(result.get(slot) or {}),
                        tool: {**tool_opts, "params": {**params, **tool_opts["params"]}},
                    },
                }
    return result


BC_PAR_PREFIX = "BC_PAR_"


def read_bc_pars(opts: Opts, env: dict[str, str | None] | None = None) -> Opts:
    """Override params from ``BC_PAR_*`` environment variables."""
    env_map = os.environ if env is None else env
    from_env: dict[str, str] = {}
    for k, v in env_map.items():
        if k.startswith(BC_PAR_PREFIX) and v is not None:
            key = k[len(BC_PAR_PREFIX) :].lower().replace("_", "-").replace(".", "-")
            from_env[key] = v
    return {**opts, "params": {**(opts.get("params") or {}), **from_env}}


@dataclass(frozen=True)
class PipelineEntry:
    name: str
    fn: Callable[[list[StepFn], Opts], Opts]
    args: str
    opts_fn: Callable[[Opts], Opts] | None = None


def to_comp_workflow(*, first_step: str, pipeline: list[PipelineEntry], last_step: str | None = None):
    """Create a composite workflow from a pipeline of tool workflows."""

    def comp(step_fns: list[StepFn], opts: Opts) -> Opts:
        last = last_step or f"{first_step.split('/')[0]}/end"
        globals_opts = new_prefix(select_globals(opts), first_step)
        step_info: dict[str, dict[str, Any]] = {}
        for entry in pipeline:
            parsed = parse_args(entry.args)
            override = opts.get(f"{entry.name}-opts") or {}
            step_info[entry.name] = {
                "stepOpts": {**parsed, **globals_opts, **override},
                "optsFn": entry.opts_fn or (lambda o: o),
            }
        order = [first_step, *[e.name for e in pipeline], last]
        next_by_step = {order[i]: order[i + 1] for i in range(len(order) - 1)}
        fn_by_step: dict[str, WfStep] = {first_step: ok, last: lambda o: o}
        for e in pipeline:
            fn_by_step[e.name] = lambda o, e=e: e.fn(step_fns, o)
        steps_set = {e.name for e in pipeline}
        cur: Opts = dict(opts)

        def wire(step: str, _sfns: list[StepFn]) -> tuple[WfStep, str | None]:
            return fn_by_step.get(step, lambda o: o), next_by_step.get(step)

        def next_fn(step: str, next_step: str | None, o: Opts) -> tuple[str | None, Opts]:
            nonlocal cur
            if step in steps_set:
                cur = {**cur, "exit": o.get("exit"), "err": o.get("err"), step: o}
            else:
                cur = o
            if step == last:
                return None, cur
            if isinstance(o.get("exit"), int) and o["exit"] > 0:
                return last, cur
            info = step_info.get(next_step) if next_step else None
            if info:
                return next_step, info["optsFn"](info["stepOpts"])
            return next_step, cur

        wf = to_workflow_star(first_step=first_step, last_step=last, wire_fn=wire, next_fn=next_fn)
        return wf(step_fns, opts)

    return comp


printStepFn = print_step_fn
selectGlobals = select_globals
runSteps = run_steps
parseArgs = parse_args
newPrefix = new_prefix
mergeParams = merge_params
readBcPars = read_bc_pars
toCompWorkflow = to_comp_workflow
